File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.cpp
class Solution {
public:
    int myAtoi(std::string s) {
        std::unordered_map<char, int> digits = {
            {'0', 0}, {'1', 1}, {'2', 2}, {'3', 3}, {'4', 4},
            {'5', 5}, {'6', 6}, {'7', 7}, {'8', 8}, {'9', 9}
        };

        int flag = 1, i = 0, n = s.size(), answer = 0;

        while (i < n && s[i] == ' ') i++;

        if (i < n && (s[i] == '-' || s[i] == '+')) {
            if (s[i] == '-') flag = -1;
            i++;
        }

        while (i < n && std::isdigit(s[i])) {
            int digit = digits[s[i]];

            if (answer > (INT_MAX - digit) / 10) 
                return flag == 1 ? INT_MAX : INT_MIN;

            answer = answer * 10 + digit;
            i++;
        }

        return answer * flag;
    }
};

File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.java
class Solution {
    public int myAtoi(String s) {
        s = s.trim(); // Remove leading/trailing whitespace
        if (s.length() == 0) return 0;

        int sign = 1;
        if (s.charAt(0) == '+' || s.charAt(0) == '-') {
            sign = s.charAt(0) == '-' ? -1 : 1;
            s = s.substring(1); 
        }

        return helper(s, 0, 0L, sign);
    }

    private int helper(String s, int index, long result, int sign) {
        if (index == s.length() || !Character.isDigit(s.charAt(index))) {
            return (int)(sign * result);
        }

        int digit = s.charAt(index) - '0';
        result = result * 10 + digit;

        long signedResult = result * sign;
        if (signedResult > Integer.MAX_VALUE) return Integer.MAX_VALUE;
        if (signedResult < Integer.MIN_VALUE) return Integer.MIN_VALUE;

        return helper(s, index + 1, result, sign);
    }
}
